transformer_engine/pytorch/float8_tensor.py

```diff
@@ -435,80 +463,50 @@ def expand_as(self, other: torch.Tensor):
             return _IdentityFunc.apply(self)
         return super().expand_as(other)
 
-    def transpose(
+    def transpose_2d(
         self,
-        dim0: int = 0,
-        dim1: int = 1,
         *,
-        update_cache: str | bool = "reuse_only",
+        cache: bool = False,
+        noop_flag: Optional[torch.Tensor] = None,
     ) -> torch.Tensor:
         """
-        Swap tensor dimensions
-
-        For basic 2D matrix transposes, an optimized transpose kernel
-        is applied and a Float8Tensor is returned.
+        2D transpose with caching support.
 
         Parameters
         ----------
-        dim0: int, default = 0
-              The first dimension to be transposed
-        dim1: int, default = 1
-              The second dimension to be transposed
-        update_cache: str or bool, default = "reuse_only"
-                      Memoization behavior. Options are
-                      "reuse_only"/`False` (reuse cached value if
-                      available, otherwise calculate transpose without
-                      caching), "force"/`True` (calculate transpose
-                      and cache), "lazy" (reuse cached value if
-                      available, otherwise calculate transpose and
-                      cache if possible). Caching is only supported
-                      for basic 2D transposes and the cache is reset
-                      after any in-place operations.
-
+        cache: bool, default = `False`
+               Whether or not to cache the transpose.
+        noop_flag: Optional[torch.Tensor], default = `None`
+                   Only used if argument `cache` is `True`, ignored otherwise.
+                   A single element fp32 tensor with a value of 1.0 or 0.0
+                   which is treated as a boolean. `1.0` forces recompute
+                   and `0.0` executes a noop using the same kernel.
         """
+        assert self.dim() == 2, f"{self.dim()}-D transpose not supported."
 
-        # Check caching mode
-        if not isinstance(update_cache, str):
-            update_cache = "force" if update_cache else "reuse_only"
-        if update_cache not in ("force", "reuse_only", "lazy"):
-            raise ValueError(
-                "Supported values for update_cache are "
-                '"force" (True), "reuse_only" (False), "lazy" '
-                f"(got {update_cache})"
-            )
+        # Case: no caching.
+        if not cache:
+            return tex.fp8_transpose(self._data, self._fp8_dtype)
```

Comment:
```suggestion
        if not cache:
            if not self._transpose_invalid and self._transpose is not None:
                return self._transpose
            else:
                return tex.fp8_transpose(self._data, self._fp8_dtype)
```

Lazy transpose caching is important for https://github.com/NVIDIA/TransformerEngine/pull/707. If we want to control caching with a single kwarg, I propose the following behavior:

|   | Invalid cache | Valid cache |
|---|---|---|
| `cache=False` | Compute transpose | Return cache |
| `cache=True` | Compute transpose and store in cache | Return cache |

If we have two kwargs:

|   |   | Invalid cache | Valid cache |
|---|---|---|---|
| `force_compute=False` | `fill_cache=False` | Compute transpose | Return cache |
| `force_compute=True` | `fill_cache=False` | Compute transpose | Compute transpose and invalidate cache |
| `force_compute=False` | `fill_cache=True` | Compute transpose and store in cache | Return cache |
| `force_compute=True` | `fill_cache=True` | Compute transpose and store in cache | Compute transpose and store in cache |

See the proposed API in https://github.com/NVIDIA/TransformerEngine/pull/735.

Reply:
Could you check if the current behavior is in line with this?

The comments delineate the logic fairly well and it should support both in module transpose calculation in TE and when caching transposes outside of TE. This coupled with the `fp8_weight_caching` argument in the graphs case should handle the given scenarios. @timmoon10

Reply:
The current behavior in this PR:
|   | Invalid cache | Valid cache |
|---|---|---|
| `cache=False` | Compute transpose | Compute transpose |
| `cache=True` | Compute transpose and store in cache | Return cache |
It does not handle lazy caching.

Reply:
Currently, `cache=True` guarantees that the `Float8Tensor` has a valid cache at the end. This isn't good for some current use-cases like memory-constrained finetuning, where we're willing to accept an extra transpose kernel in order to save the memory cost of caching the transpose.